pkg/node/client.go

```diff
@@ -25,11 +25,12 @@ func NewNodeClient(identity *provider.FullIdentity, self pb.Node, dht dht.DHT) (
 		dht:   dht,
 		self:  self,
 		tc:    client,
-		cache: pool.NewConnectionPool(),
+		cache: *pool.NewConnectionPool(),
```

Comment:
Connection Pool has a mutex, so this might cause problems in the future. Rather add an init function to the `ConnectionPool` that initializes it in place.

```
node := &Node{ ... }
node.cache.Init()
return node, nil
```

Reply:
Can you explain this further? Thanks :)

Reply:
Could you explain this further? I'm not sure I understand

Reply:
I'm missing the issue here as well, could you explain @egonelbre ?

Reply:
Mutexes are non-copyable. So something like:

```
var a sync.Mutex
b := a
```

Is incorrect code, since the correctness of mutex depends on the content being synchronized between goroutines, so, making a copy would break it.

This of course also happens when you copy a struct which contains a mutex field.
```
type X struct {
    mu sync.Mutex
}

var a X
b := a
```

Init avoids this because the you would be updating the content inplace rather than copying it.

Now depending on the exact situation it might be fine (like a very simple constructor).